### PR TITLE
Better: Add Topology endpoint to API doc. RD-21094

### PR DESCRIPTION
**Pull request by ylecuyer**
Adding the topology endpoint to the doc, see inline comments for details

### CODE REVIEW

File: specs/engage-digital_openapi3.yaml [hunk 1]
```diff
@@ -97,6 +98,138 @@ tags:
...
+        '200':
+            description: Created topology 
+            content:
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/Topology'
+      summary: Creating a topology
+      tags:
+        - Topologies
+    get:
```
Comment (ylecuyer) [inline specs/engage-digital_openapi3.yaml:217]: Screenshot_2022-02-01_16-23-05

Tested with:

```
ruby -Ilib -rswagger_client -rawesome_print -e "SwaggerClient.configure { |c| c.host = 'domain-test.api.engagement.dimelo.info'; c.base_path = '1.0'}; ap SwaggerClient::TopologiesApi.new.topologies_get(query_params: { access_token: '****' })"

{
      :count => 7,
      :limit => 30,
     :offset => 0,
    :records => [
        [0] {
                    :id => "56570a4e7765620a6c0006e9",
            :created_at => "2015-11-26T13:34:06Z",
            :updated_at => "2021-09-02T13:46:19Z",
                :config => {
                :inputs => {
                       :inbound => "import_step",
                    :transfered => "transfer_step"
                },
                 :steps => [
                    [ 0] {
                        :links => {
                               :async => "async_intervention_target",
                            :realtime => "realtime_intervention_target"
                        },
                        :_type => "Routing::ImportStep",
                        :label => "import_step"
                    },
                    [ 1] {
                                            :links => {
                            :failure => "thread_target",
                            :reroute => "async_intervention_target"
                        },
                                             :name => "",
                                  :timeout_seconds => 2592000,
                                            :_type => "Routing::InterventionAgentsTarget",
                                            :label => "async_intervention_target",
                                       :task_class => "ProcessingTask",
                        :timeout_in_business_hours => true,
                             :check_agent_presence => false
                    },
...
```

File: specs/engage-digital_openapi3.yaml [hunk 2]
```diff
@@ -97,6 +98,138 @@ tags:
...
+        '200':
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Topology'
+          description: Success
+      summary: Deleting a topology
+      tags:
+        - Topologies
+    get:
```
Comment (ylecuyer) [inline specs/engage-digital_openapi3.yaml:172]: Screenshot_2022-02-01_16-28-25

Tested with:

```
ruby -Ilib -rswagger_client -rawesome_print -e "SwaggerClient.configure { |c| c.host = 'host'; c.base_path = '1.0'}; ap SwaggerClient::TopologiesApi.new.get_topology('56570a4e7765620a6c0006e9', query_params: { access_token: '***' })"

{
          :name => "Default Topology",
    :created_at => "2015-11-26T13:34:06Z",
    :updated_at => "2021-09-02T13:46:19Z",
            :id => "56570a4e7765620a6c0006e9",
        :config => {
        :inputs => {
               :inbound => "import_step",
            :transfered => "transfer_step"
        },
         :steps => [
            [ 0] {
                :links => {
                       :async => "async_intervention_target",
                    :realtime => "realtime_intervention_target"
                },
                :_type => "Routing::ImportStep",
                :label => "import_step"
            },
            [ 1] {
                                    :links => {
                    :failure => "thread_target",
                    :reroute => "async_intervention_target"
                },
                                     :name => "",
                          :timeout_seconds => 2592000,
                                    :_type => "Routing::InterventionAgentsTarget",
                                    :label => "async_intervention_target",
                               :task_class => "ProcessingTask",
                :timeout_in_business_hours => true,
                     :check_agent_presence => false
            },
...
```

File: specs/engage-digital_openapi3.yaml [hunk 3]
```diff
@@ -97,6 +98,138 @@ tags:
   - name: Surveys
 
 paths:
+  '/topologies/{topologyId}/activate':
+    put:
```
Comment (ylecuyer) [inline specs/engage-digital_openapi3.yaml:102]: Screenshot_2022-02-01_16-32-22

Tested with:

```
ruby -Ilib -rswagger_client -rawesome_print -e "SwaggerClient.configure { |c| c.host = 'host'; c.base_path = '1.0'}; ap SwaggerClient::TopologiesApi.new.activate_topology('56570a4e7765620a6c0006e9', query_params: { access_token: '***' })"

{
          :name => "Default Topology",
    :created_at => "2015-11-26T13:34:06Z",
    :updated_at => "2022-02-01T15:30:53Z",
            :id => "56570a4e7765620a6c0006e9",
        :config => {
        :inputs => {
               :inbound => "import_step",
            :transfered => "transfer_step"
        },
         :steps => [
            [ 0] {
                :links => {
                       :async => "async_intervention_target",
                    :realtime => "realtime_intervention_target"
                },
                :_type => "Routing::ImportStep",
                :label => "import_step"
            },
            [ 1] {
                                    :links => {
                    :failure => "thread_target",
                    :reroute => "async_intervention_target"
                },
                                     :name => "",
                          :timeout_seconds => 2592000,
                                    :_type => "Routing::InterventionAgentsTarget",
                                    :label => "async_intervention_target",
                               :task_class => "ProcessingTask",
                :timeout_in_business_hours => true,
                     :check_agent_presence => false
            },
...
```

File: specs/engage-digital_openapi3.yaml [hunk 4]
```diff
@@ -97,6 +98,138 @@ tags:
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Topology'
+          description: Success
+      summary: Activating a topology
+      tags:
+        - Topologies
+  '/topologies/{topologyId}':
+    put:
```
Comment (ylecuyer) [inline specs/engage-digital_openapi3.yaml:126]: Screenshot_2022-02-01_16-55-15

Tested with:

```
ruby -Ilib -rswagger_client -rawesome_print -e "SwaggerClient.configure { |c| c.host = 'host'; c.base_path = '1.0'}; ap SwaggerClient::TopologiesApi.new.update_topology('56570a4e7765620a6c0006e9', name: 'test from API', query_params: { access_token: '***' })"

{
          :name => "test from API",
    :created_at => "2015-11-26T13:34:06Z",
    :updated_at => "2022-02-01T15:54:01Z",
            :id => "56570a4e7765620a6c0006e9",
        :config => {
        :inputs => {
               :inbound => "import_step",
            :transfered => "transfer_step"
        },
         :steps => [
            [ 0] {
                :links => {
                       :async => "async_intervention_target",
                    :realtime => "realtime_intervention_target"
                },
                :_type => "Routing::ImportStep",
                :label => "import_step"
            },
            [ 1] {
                                    :links => {
                    :failure => "thread_target",
                    :reroute => "async_intervention_target"
                },
                                     :name => "",
                          :timeout_seconds => 2592000,
                                    :_type => "Routing::InterventionAgentsTarget",
                                    :label => "async_intervention_target",
                               :task_class => "ProcessingTask",
                :timeout_in_business_hours => true,
                     :check_agent_presence => false
...
```

File: specs/engage-digital_openapi3.yaml [hunk 5]
```diff
@@ -97,6 +98,138 @@ tags:
...
+        '200':
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Topology'
+          description: Success
+      summary: Getting a topology from its id
+      tags:
+        - Topologies
+  /topologies:
```
Comment (ylecuyer) [inline specs/engage-digital_openapi3.yaml:200]: Screenshot_2022-02-01_17-02-39

Tested with:

```
ruby -Ilib -rswagger_client -rawesome_print -e "SwaggerClient.configure { |c| c.host = 'host'; c.base_path = '1.0'}; ap SwaggerClient::TopologiesApi.new.topologies_post('create from api', '{  \"inputs\": {    \"inbound\": \"import_step\",    \"transfered\": \"transfer_step\"  },  \"steps\": [    {      \"links\": {        \"async\": \"triage\",        \"realtime\": \"triage\"      },      \"_type\": \"Routing::ImportStep\",      \"label\": \"import_step\"    },    {      \"links\": {        \"async\": \"triage\",        \"realtime\": \"triage\"      },      \"_type\": \"Routing::TransferStep\",      \"label\": \"transfer_step\"    },    {      \"links\": {      },      \"_type\": \"Routing::Enqueue\",      \"label\": \"triage\",      \"queue\": \"triage\"    }  ]}', query_params: { access_token: '***' })"

{
          :name => "create from api",
    :created_at => "2022-02-01T16:02:13Z",
    :updated_at => "2022-02-01T16:02:13Z",
            :id => "61f959856c98bd5a2a8ec15d",
        :config => {
        :inputs => {
               :inbound => "import_step",
            :transfered => "transfer_step"
        },
         :steps => [
            [0] {
                :links => {
                       :async => "triage",
                    :realtime => "triage"
                },
                :_type => "Routing::ImportStep",
                :label => "import_step"
            },
            [1] {
                :links => {
                       :async => "triage",
                    :realtime => "triage"
                },
                :_type => "Routing::TransferStep",
                :label => "transfer_step"
            },
            [2] {
                :links => {},
                :_type => "Routing::Enqueue",
                :label => "triage",
                :queue => "triage"
            }
        ]
    }
}
```

**Comment by CedricBm** [comment]
Since you're adding new endpoints, can you update the postman collection too please? cf https://github.com/ringcentral/engage-digital-api-docs/tree/master/specs#usage

File: specs/engage-digital_openapi3.yaml [hunk 6]
```diff
@@ -97,6 +98,143 @@ tags:
   - name: Surveys
 
 paths:
+  '/topologies/{topologyId}/activate':
+    put:
+      description: >-
+        This method activates an existing topology from given attributes and renders it in case of success.
+        
+        
+        Authorization: only users that are techteam.
```
Comment (CedricBm) [inline specs/engage-digital_openapi3.yaml:107]: Arf, actually this isn't true anymore. The GDoc is not up to date. Everyone having the "manage topology" permission can do it.

Same for the other endpoints, none are techteam only anymore.

File: specs/engage-digital_openapi3.yaml [hunk 7]
```diff
@@ -97,6 +98,143 @@ tags:
...
+          description: Success
+      summary: Getting a topology from its id
+      tags:
+        - Topologies
+  /topologies:
+    post:
+      description: >-
+        This method creates a topology. In case of success it renders the topology, otherwise, it renders an error (422 HTTP code).
+      
+          Authorization: only users that are superadmin.
```
Comment (CedricBm) [inline specs/engage-digital_openapi3.yaml:205]: Indentation and permission change

```suggestion
        Authorization: only users that have the right to manage topologies.
```

**Comment by ylecuyer** [comment]
> Since you're adding new endpoints, can you update the postman collection too please? cf https://github.com/ringcentral/engage-digital-api-docs/tree/master/specs#usage

done

**Comment by CedricBm** [comment]
LGTM. We'll be able to deprecate the GDoc once this will be deployed. Supposedly end of Feb

**Review by kalilz4485 — APPROVED** [review]
Looks good

**Comment by CoCraig** [comment]
@ylecuyer I think you forgot to add Topologies to the x-tag-groups. It's not showing up in the API Reference in the Developer Portal.  I believe it should be under "Provisioning"
  - name: Provisioning
    tags:
      - Categories
      - CustomFields
      - Locales
      - Reply Assistant Entries
      - Reply Assistant Groups
      - Reply Assistant Versions
      - Roles
      - Settings
      - Tags
      - Teams
      - Time Sheets
      - Timezones
      {Maybe add Topologies here?}
      - Users
      - User Capacities
      - User Signatures

**Comment by ylecuyer** [comment]
Thanks for the info, I'll open another PR to add it to x-tag-group though I think it should be in the Routing section.

Is there a way to generate a demo page of the website? afaict the mkdoc server only handle the guide part but there is nothing for the reference part.

Also should we add some test to make sure all tags are in x-tag-group to avoid this issue in the future?

**Comment by CoCraig** [comment]
> Is there a way to generate a demo page of the website? afaict the mkdoc server only handle the guide part but there is nothing for the reference part.

Unfortunately there isn't.  A test would be good to add, maybe to a GitHub action.

**Comment by ylecuyer** [comment]
> > Is there a way to generate a demo page of the website? afaict the mkdoc server only handle the guide part but there is nothing for the reference part.
> 
> Unfortunately there isn't. A test would be good to add, maybe to a GitHub action.

Done in https://github.com/ringcentral/engage-digital-api-docs/pull/76

**Pull request merged**